src/panoptes/pocs/focuser/birger.py

```diff
@@ -249,13 +249,15 @@ def _send_command(self, command, response_length=None):
         # Clear the input buffer in case there's anything left over in there.
         self._serial_port.reset_input_buffer()
 
-        # Send command
-        self._serial_io.write(command + '\r')
-
         # In verbose mode adaptor will first echo the command
-        echo = self._serial_io.readline().rstrip()
-        assert echo == command, self.logger.warning("echo != command: {} != {}".format(
-            echo, command))
+        echo = [self._serial_io.readline().rstrip()]
+        while echo[0] != command:
+            # Send command
+            self._serial_io.write(command + '\r')
+            echo[0] = self._serial_io.readline().rstrip()
+
+        assert echo[0] == command, self.logger.warning("echo != command: {} != {}".format(
+            echo[0], command))
 
         # Adaptor should then send 'OK', even if there was an error.
         ok = self._serial_io.readline().rstrip()
```

Comment:
Note that there are several more steps to verifying responses from the Birger. Checking that the Birger echos the command that was sent to it is just the first step.

We also check here that this is followed by 'OK', check for error codes, then retrieve any additional lines of response that are expected for the given command type.

The serial comms problems are just as likely to effect these subsequent checks as the initial echo check, so they should all be moved inside the retry loop.

Reply:
But what if the response from the birger is correct when the command is sent for the first time as expected? If everything is inside the retry loop (which only runs if `echo != command`), that'd mean there would not be subsequent checks if the loop is not triggered.